### PR TITLE
priority_auth_renewer: only try to republish if failed to do so

### PR DESCRIPTION
**Pull request by unleashed**
There is no longer a race condition here, so while it is good to
retry, it is not any longer because of the client requesting the
combination before it actually subscribed to the response channel.

Now we just keep retrying for a while if publishing the results
fail.

### CODE REVIEW

**Comment by davidor** [comment]
For reference, this is the PR in apicast-xc that solves the race condition: https://github.com/3scale/apicast-xc/pull/22

**Comment by davidor** [comment]
@unleashed , we need to delete the comment in `publish_auth_repeatedly` that explains the old race condition.

File: lib/xcflushd/priority_auth_renewer.rb [hunk 1]
```diff
@@ -195,9 +195,12 @@ def publish_auth_repeatedly(combination, authorization)
       # incoming requests slow because tasks accumulate in the thread pool
       # waiting.
       publish_failures = 0
+      successful = false
       PUBLISH_WAIT_TIMES.each do |wait_time|
         begin
           publish_auth(combination, authorization)
+          successful = true
+          break
```
Comment (unleashed) [inline lib/xcflushd/priority_auth_renewer.rb]: I am going to write this a bit more correctly.

**Comment by unleashed** [comment]
I have removed the comment and moved the "repeatedly" part into `publish_auth`.

File: lib/xcflushd/priority_auth_renewer.rb [hunk 2]
```diff
@@ -173,53 +173,33 @@ def channel_for_combination(combination)
...
-      # I cannot think of an easy way to solve this. There is some time
-      # between the moment the requests performs the publish and the
-      # subscribe actions. To mitigate the problem we can publish several
-      # times during some ms. We will see if this is good enough.
-      # Trade-off: Waiting long times between publishing attempts reduces the
-      # probability of triggering the problem described. However, it also makes
-      # incoming requests slow because tasks accumulate in the thread pool
-      # waiting.
+    def publish_auth(combination, authorization)
+      msg = if authorization.authorized?
```
Comment (davidor) [inline lib/xcflushd/priority_auth_renewer.rb]: I see that you changed this so it's calculated only once :+1: 
How about extracting this to its own method? something like `msg_to_publish(authorization)` ?

File: lib/xcflushd/priority_auth_renewer.rb [hunk 3]
```diff
@@ -173,53 +173,33 @@ def channel_for_combination(combination)
...
+          successful = true
+          break
         end
         sleep(wait_time)
       end
 
       if publish_failures > 0
         logger.warn('There was an error while publishing a response in the '\
-                    "priority channel. Combination: #{combination}".freeze)
+                    "priority channel. Combination: #{combination} - Status: #{successful ? 'successfully published'.freeze : 'could not publish'.freeze}")
```
Comment (davidor) [inline lib/xcflushd/priority_auth_renewer.rb]: I'd changed the if condition to `unless successful`

File: lib/xcflushd/priority_auth_renewer.rb (hunk 2 above)
Comment (unleashed) [inline lib/xcflushd/priority_auth_renewer.rb]: IMO there is a more important and interesting way to split this, which would be to provide a `publish_message` with error handling that would receive the message, and `publish_auth` would call it - I'd rather prefer extracting that. Once that is done I think the method should be simple enough. I will give it a go.

File: lib/xcflushd/priority_auth_renewer.rb (hunk 3 above)
Comment (davidor) [inline lib/xcflushd/priority_auth_renewer.rb]: Also, minor thing, this line is too long. Can you split it?

File: lib/xcflushd/priority_auth_renewer.rb (hunk 2 above)
Comment (unleashed) [inline lib/xcflushd/priority_auth_renewer.rb]: Actually I have ended up extracting both. Check it out in a couple minutes.

File: lib/xcflushd/priority_auth_renewer.rb [hunk 4]
```diff
@@ -173,51 +173,44 @@ def channel_for_combination(combination)
...
-               authorization.reason ? "0:#{authorization.reason}" : '0'.freeze
-            end
-
-      redis_pub.publish(channel_for_combination(combination), msg)
+      success, failures =
+        publish_message(channel_for_combination(combination),
+                        authorization_message(authorization))
+
+      if failures > 0
+        logger.warn("There was an error while publishing a response in the " \
```
Comment (davidor) [inline lib/xcflushd/priority_auth_renewer.rb]: I wouldn't log anything if at least one attempt succeeded.

**Comment by davidor** [comment]
This looks good but there's a missing test case. When there's an error publishing a response, the priority auth tries a few more times. We have a test case for the scenario where all the attempts fail, but not for the case where some attempts fail and one is successful.

See https://github.com/3scale/xcflushd/blob/master/spec/priority_auth_renewer_spec.rb#L165

**Comment by davidor** [comment]
@unleashed  I added the test case that was missing. Let me know what you think.

**Comment by unleashed** [comment]
Looks good. Is the rest ok?

**Pull request merged**